### ISSUE TITLE
Segfault when accessing a group or environment in python

### ISSUE THREAD

**Issue by pkratoch** [rpm-software-management/dnf5#1530]
Hi, I need help with this issue with comps in python when they are created in another function. This is the reproducer:

```
>>> import libdnf5
>>> base = libdnf5.base.Base()
>>> base.setup()
>>> base.get_repo_sack().load_repos(True)
>>> def get_env(base):
...     environments = libdnf5.comps.EnvironmentQuery(base)
...     return next(iter(environments))
... 
>>> get_env(base).get_environmentid()
Segmentation fault (core dumped)
```

This works:

```
>>> import libdnf5
>>> base = libdnf5.base.Base()
>>> base.setup()
>>> base.get_repo_sack().load_repos(True)
>>> environments = libdnf5.comps.EnvironmentQuery(base)
>>> env = next(iter(environments))
>>> env.get_environmentid()
```

I would think there is something wrong with the copy constructor / assignment operator, but there is no problem e.g. for advisories that have basically the same copy constructor.
Is there something obvious I am missing?

**Comment by jan-kolarik** [comment]
Hmm, my first though would be that the environments list is destroyed when going out of the scope from the `get_env()` function. But if the same code works for AdvisoryQuery, we would need to dig deeper I guess ...

**Comment by kontura** [comment]
> Hmm, my first thought would be that the environments list is destroyed when going out of the scope from the `get_env()` function. But if the same code works for AdvisoryQuery, we would need to dig deeper I guess ...

I think that is it. `AdvisoryQuery` works quite a bit differently (more like `PackageQuery`), it is just a `SolvMap` so it doesn't store the advisory objects.

On the other hand `EnvironmentQuery`, `GroupQuery` and `RepoQuery` ultimately inherit from `libdnf5::set` template which stores data in a `std::set<T>`. When it goes out of scope the set destroys its content.

The reproducer doesn't crash with `RepoQuery` because [it doesn't store the `Repo` objects in the query but rather a `RepoWeakPtr`](https://github.com/rpm-software-management/dnf5/blob/main/include/libdnf5/repo/repo_query.hpp#L44).

So to fix this I think the comps queries would have to also use a weak pointer [here](https://github.com/rpm-software-management/dnf5/blob/main/include/libdnf5/comps/environment/query.hpp#L34) instead of the actual objects.

**Comment by j-mracek** [comment]
May be this is missing what we have for `packageset`
```
%rename(next) libdnf5::rpm::PackageSetIterator::operator++();
%rename(value) libdnf5::rpm::PackageSetIterator::operator*();
```

**Comment by pkratoch** [comment]
> May be this is missing what we have for `packageset`
> 
> ```
> %rename(next) libdnf5::rpm::PackageSetIterator::operator++();
> %rename(value) libdnf5::rpm::PackageSetIterator::operator*();
> ```

This would mean creating GroupSet and GroupSetIterator and the same for environments? At the first glance, it looks like a more complicated approach.

But anyway, thank you all for great advice, I'll look into both suggestions.

**Comment by j-mracek** [comment]
I agree with @kontura. The problem is that iterator returns pointer of object and owner of the object is comps sack.
```
>>> import libdnf5
>>> base = libdnf5.base.Base()
>>> base.setup()
>>> base.get_repo_sack().load_repos(True)
>>> def get_env(base):
...     environments = libdnf5.comps.EnvironmentQuery(base)
...     return libdnf5.comps.Environment(next(iter(environments)))
... 
>>> get_env(base).get_environmentid()
```

I am recommending to unify implementation with Repo Sack, but it might be painful.

**Comment by pkratoch** [comment]
Fixed by https://github.com/rpm-software-management/dnf5/pull/2351 (I couldn't link the issue from the PR for some reason...)
